lib/libwebsockets.c

```diff
@@ -510,14 +510,28 @@ lws_close_free_wsi(struct lws *wsi, enum lws_close_status reason)
 	    !wsi->socket_is_permanently_unusable) {
 #ifdef LWS_OPENSSL_SUPPORT
 		if (lws_is_ssl(wsi) && wsi->ssl)
-               {
-                       lwsl_info("%s: shutting down SSL connection: %p (ssl %p, sock %d, state %d)\n", __func__, wsi, wsi->ssl, (int)(long)wsi->desc.sockfd, wsi->state);
+		{
+			lwsl_info("%s: shutting down SSL connection: %p (ssl %p, sock %d, state %d)\n", __func__, wsi, wsi->ssl, (int)(long)wsi->desc.sockfd, wsi->state);
 			n = SSL_shutdown(wsi->ssl);
-                       if (n == 1) /* If finished the SSL shutdown, then do socket shutdown, else need to retry SSL shutdown */
-                               n = shutdown(wsi->desc.sockfd, SHUT_WR);
-                       else
-                               lws_change_pollfd(wsi, LWS_POLLOUT, LWS_POLLIN);
-               }
+			if (n == 1) /* If finished the SSL shutdown, then do socket shutdown, else need to retry SSL shutdown */
+				n = shutdown(wsi->desc.sockfd, SHUT_WR);
+			else if (n == 0)
+				lws_change_pollfd(wsi, LWS_POLLOUT, LWS_POLLIN);
```

Comment:
When we call `SSL_shutdown` and the other end hasn't sent an SSL close message, we will hit the `n == 0` case, so we are now in the state of having sent an SSL close message and we are giving the other side a chance to acknowledge the close with their own SSL close. This is why we start waiting for the reply, we can't do the second `SSL_shutdown` call in the `n == 0` case here, as the sockets are non-blocking, so there won't have been time for the message to be sent and a reply received.
When the reply comes in, this will trigger the code in `lws_service_fd_tsi` to do the next `SSL_shutdown` call which should then return 1.